File: src/java/foundgine-runtime/src/main/java/com/foundgine/runtime/FoundgineServiceCollectionExtensions.java
package com.foundgine.runtime;

import com.foundgine.core.execution.*;
import com.foundgine.core.semantic.SemanticContractSnapshot;
import com.foundgine.core.semantic.SemanticModel;
import com.foundgine.core.semantic.authorization.AllowAllSemanticAuthorizationPolicy;
import com.foundgine.core.semantic.authorization.ISemanticAuthorizationPolicy;
import com.foundgine.core.semantic.security.SecurityInvariantContractValidator;
import java.util.Objects;

/**
 * Framework-neutral Java counterpart of the C# {@code AddFoundgine}
 * extension methods.
 *
 * <p>The C# API targets {@code IServiceCollection}. Java deliberately does
 * not introduce a dependency on one particular DI framework. This adapter
 * performs the same startup trust-boundary work and registers the resulting
 * objects in {@link FoundgineServiceRegistry}. Applications can then bridge
 * that registry into their DI container.
 */
public final class FoundgineServiceCollectionExtensions {
    private FoundgineServiceCollectionExtensions() {}

    public static FoundgineServiceRegistry addFoundgine(
            FoundgineServiceRegistry services,
            SemanticModel model,
            IProviderPlanCompiler compiler,
            IExecutionProvider provider) {
        return addFoundgine(services, model, new AllowAllSemanticAuthorizationPolicy(), compiler, provider);
    }

    public static FoundgineServiceRegistry addFoundgine(
            FoundgineServiceRegistry services,
            SemanticModel model,
            ISemanticAuthorizationPolicy authorizationPolicy,
            IProviderPlanCompiler compiler,
            IExecutionProvider provider) {
        Objects.requireNonNull(services, "services");
        Objects.requireNonNull(model, "model");
        Objects.requireNonNull(authorizationPolicy, "authorizationPolicy");
        Objects.requireNonNull(compiler, "compiler");
        Objects.requireNonNull(provider, "provider");

        FoundgineOptions options = new FoundgineOptions()
                .model(model)
                .authorizationPolicy(authorizationPolicy);

        SemanticContractSnapshot contract = model.freeze().createSnapshot();
        SecurityInvariantContractValidator.ensureContractValid(
                com.foundgine.core.semantic.capabilities.SemanticCapabilityContractDiscovery
                        .describe(model, authorizationPolicy));

        services.addSingletonInstance(FoundgineOptions.class, options);
        services.addSingletonInstance(SemanticContractSnapshot.class, contract);
        services.addSingleton(FoundgineEngine.class,
                () -> new FoundgineEngine(options, contract, compiler, provider));
        services.addSingleton(IFoundgine.class, () -> services.getRequiredService(FoundgineEngine.class));
        services.addSingleton(IFoundgineExecutor.class, () -> services.getRequiredService(FoundgineEngine.class));

        if (options.mutationSchema() != null && options.mutationProvider() != null) {
            services.addSingleton(IFoundgineMutations.class, () -> new FoundgineMutationEngine(
                    options.mutationSchema(),
                    options.authorizationPolicy(),
                    options.mutationProvider(),
                    options.model(),
                    options.warrantKeyResolver(),
                    options.expectedWarrantIssuer(),
                    options.warrantReplayStore(),
                    options.securityResourceLimits()));
        }
        return services;
    }

    public static FoundgineServiceRegistry addFoundgine(
            FoundgineServiceRegistry services,
            FoundgineOptions options,
            IProviderPlanCompiler compiler,
            IExecutionProvider provider) {
        Objects.requireNonNull(options, "options");
        if (options.model() == null)
            throw new IllegalStateException("Foundgine requires FoundgineOptions.model().");
        if (options.authorizationPolicy() == null) {
            if (options.authorizationConfiguration() != null)
                options.authorizationPolicy(new com.foundgine.core.semantic.authorization.ConfiguredSemanticAuthorizationPolicy(
                        options.authorizationConfiguration(), options.authorizationContext()));
            else
                throw new IllegalStateException(
                        "Foundgine requires an ISemanticAuthorizationPolicy or configured authorization.");
        }
        Objects.requireNonNull(services, "services");
        Objects.requireNonNull(compiler, "compiler");
        Objects.requireNonNull(provider, "provider");

        SemanticContractSnapshot contract = options.model().freeze().createSnapshot();
        services.addSingletonInstance(FoundgineOptions.class, options);
        services.addSingletonInstance(SemanticContractSnapshot.class, contract);
        services.addSingleton(FoundgineEngine.class,
                () -> new FoundgineEngine(options, contract, compiler, provider));
        services.addSingleton(IFoundgine.class, () -> services.getRequiredService(FoundgineEngine.class));
        services.addSingleton(IFoundgineExecutor.class, () -> services.getRequiredService(FoundgineEngine.class));
        if (options.mutationSchema() != null && options.mutationProvider() != null) {
            services.addSingleton(IFoundgineMutations.class, () -> new FoundgineMutationEngine(
                    options.mutationSchema(), options.authorizationPolicy(), options.mutationProvider(),
                    options.model(), options.warrantKeyResolver(), options.expectedWarrantIssuer(),
                    options.warrantReplayStore(), options.securityResourceLimits()));
        }
        options.applyCapabilities(services);
        return services;
    }
}

File: src/csharp/Foundgine.Runtime/Hosting/FoundgineServiceCollectionExtensions.cs
using Foundgine.Core.Execution;
using Foundgine.Core.Semantic.Metadata;
using Foundgine.Core.Semantic;
using Foundgine.Core.Semantic.Authorization;
using Microsoft.Extensions.DependencyInjection;

namespace Foundgine.Runtime;

/// <summary>
/// Dependency-injection registration for the application-facing Foundgine API.
/// Provider adapters register IProviderPlanCompiler and IExecutionProvider.
/// </summary>
public static class FoundgineServiceCollectionExtensions
{
    /// <summary>
    /// Registers Foundgine with the supplied semantic model and the default allow-all policy.
    /// Applications that require authorization should use the overload that supplies an explicit policy.
    /// </summary>
    public static IServiceCollection AddFoundgine(
        this IServiceCollection services,
        SemanticModel model)
    {
        ArgumentNullException.ThrowIfNull(services);
        ArgumentNullException.ThrowIfNull(model);

        return services.AddFoundgine(options =>
        {
            options.Model = model;
            options.AuthorizationPolicy = new AllowAllSemanticAuthorizationPolicy();
        });
    }

    public static IServiceCollection AddFoundgine(
        this IServiceCollection services,
        SemanticModel model,
        ISemanticAuthorizationPolicy authorizationPolicy)
    {
        ArgumentNullException.ThrowIfNull(services);
        ArgumentNullException.ThrowIfNull(model);
        ArgumentNullException.ThrowIfNull(authorizationPolicy);

        return services.AddFoundgine(options =>
        {
            options.Model = model;
            options.AuthorizationPolicy = authorizationPolicy;
        });
    }

    public static IServiceCollection AddFoundgine(
        this IServiceCollection services,
        Action<FoundgineOptions> configure)
    {
        ArgumentNullException.ThrowIfNull(services);
        ArgumentNullException.ThrowIfNull(configure);

        var options = new FoundgineOptions();
        configure(options);
        options.ApplyCapabilities(services);

        if (options.Model is null && options.Metadata is not null)
        {
            var builder = options.Metadata.FromMetadata();
            options.SemanticConfiguration?.Invoke(builder);
            options.Model = builder.Build();
        }

        if (options.Model is null)
            throw new InvalidOperationException(
                "Foundgine requires a SemanticModel or structural metadata via UseMetadata().");

        // Cross the construction/trust boundary once during application startup.
        // Runtime consumers receive the immutable snapshot rather than the
        // builder/lifecycle representation.
        var contract = options.Model.Freeze().CreateSnapshot();

        if (options.AuthorizationPolicy is null && options.AuthorizationConfiguration is not null)
        {
            options.AuthorizationPolicy = new ConfiguredSemanticAuthorizationPolicy(
                options.AuthorizationConfiguration,
                options.AuthorizationContext);
        }

        if (options.AuthorizationPolicy is null)
            throw new InvalidOperationException(
                "Foundgine requires an ISemanticAuthorizationPolicy or configured authorization.");

        services.AddSingleton(options);
        services.AddSingleton(contract);
        services.AddSingleton<ISemanticContractProvider>(new SemanticContractProvider(contract));
        services.AddSingleton<FoundgineEngine>(serviceProvider => new FoundgineEngine(
            serviceProvider.GetRequiredService<FoundgineOptions>(),
            serviceProvider.GetRequiredService<SemanticContractSnapshot>(),
            serviceProvider.GetRequiredService<IProviderPlanCompiler>(),
            serviceProvider.GetRequiredService<IExecutionProvider>()));
        services.AddSingleton<IFoundgine>(serviceProvider =>
            serviceProvider.GetRequiredService<FoundgineEngine>());
        services.AddSingleton<IFoundgineExecutor>(serviceProvider =>
            serviceProvider.GetRequiredService<FoundgineEngine>());

        if (options.MutationSchema is not null && options.MutationProvider is not null)
            services.AddSingleton<IFoundgineMutations>(_ => new FoundgineMutationEngine(
                options.MutationSchema,
                options.AuthorizationPolicy,
                options.MutationProvider,
                options.Model,
                options.WarrantKeyResolver,
                options.ExpectedWarrantIssuer,
                options.WarrantReplayStore));

        return services;
    }
}